File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    const score = [];
    let num="";
    for(let i of dartResult){
        if(Number(i)%1===0){
            num+=i;
        }else if(i==="S"){
            score.push(Number(num));
            num="";
        }else if(i==="D"){
            score.push(Number(num));
            score[score.length-1]=score.at(-1)**2;
            num="";
        }else if(i==="T"){
            score.push(Number(num));
            score[score.length-1]=score.at(-1)**3;
            num="";
        }else if(i==="*"){
            score[score.length-1]=score.at(-1)*2;
            if(score.length>1){
                score[score.length-2]=score.at(-2)*2;
            }
        }else if(i==="#"){
            score[score.length-1]=score.at(-1)*(-1);
        }
    }
    const answer = score.reduce((acc,cur)=>{
            return acc+cur;
        },0)
    return answer;
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;

class Solution {
    public int solution(String dartResult) {
        int[] score = new int[3];
        
        int idx=0;
        for(int i=0; i<3; i++){
            char c = dartResult.charAt(idx);
            if(c>='0' && c<='9'){
                int n = c-'0';
                if(c=='1' && dartResult.charAt(idx+1)=='0'){
                    idx++;
                    n = 10;
                }
                char next = dartResult.charAt(++idx);
                while(next<'0' || next>'9'){
                    if(next=='S'){
                        n = n;
                    }else if(next=='D'){
                        n = n*n;
                    }else if(next=='T'){
                        n = n*n*n;
                    }else if(next=='#'){
                        n = n * (-1);
                    }else{
                        n = n * 2;
                        if(i!=0){
                            score[i-1] = score[i-1]*2;
                        }
                    }
                    if(idx==dartResult.length()-1) break;
                    next = dartResult.charAt(++idx);
                }
                score[i]=n;
            }
        }
        
        int answer = score[0]+score[1]+score[2];
        
        return answer;
    }
}